frontend/src/components/PetitionCard.tsx

```diff
@@ -0,0 +1,244 @@
+import { submitPetitionSignature } from "../utils/backend";
+import RoundedBox from "./RoundedBox";
+
+interface Props {
+  className?: string;
+  petition: PetitionDetail;
+  link?: ReactNode;
+}
+
+enum State {
+  Initial,
+  Submitting,
+  Error,
+  Existed,
+  Success,
+}
+
+const SESSION_KEY = "petitionData";
+
+function ensureSessionStorage(): { [key: string]: boolean } {
+  const data = sessionStorage.getItem(SESSION_KEY);
+  if (!data) {
+    sessionStorage.setItem(SESSION_KEY, JSON.stringify({}));
+    return {};
+  }
+  return JSON.parse(data);
+}
+
+function savePetition(petition: PetitionDetail) {
+  let data = ensureSessionStorage();
+  data[petition.id] = true;
+  sessionStorage.setItem(SESSION_KEY, JSON.stringify(data));
+}
+
+async function onSubmit(
+  petition: PetitionDetail,
+  name: string,
+  email: string,
+  questionAnswers: Array<{ question: string; answer: string }>,
+  setState: Dispatch<SetStateAction<State>>
+) {
+  setState(State.Submitting);
+  try {
+    const result = await submitPetitionSignature(
+      petition,
+      name,
+      email,
+      questionAnswers
+    );
+    if (!result) {
+      setState(State.Error);
+    }
+
+    savePetition(petition);
+    if (result.existed) {
+      setState(State.Existed);
+    } else {
+      setState(State.Success);
+    }
+  } catch {
+    setState(State.Error);
+  }
+}
+
+function renderInputForQuestion(
+  q: Question & { state: [string, Dispatch<SetStateAction<string>>] }
+) {
+  if ("choices" in q) {
+    const emptyOption = <option key="empty" value=""></option>;
+    return (
+      <select
+        className="w-full h-8 rounded-md text-black p-2"
+        onChange={(e) => q.state[1](e.target.value)}
+        required={q.required ? true : undefined}
+      >
+        {emptyOption}
+        {q.choices.map((c, i) => {
+          return (
+            <option key={i.toString()} value={c}>
+              {c}
+            </option>
+          );
+        })}
+      </select>
+    );
+  }
+  return (
+    <input
+      className="w-full h-8 rounded-md text-black p-2"
+      value={q.state[0]}
+      onInput={(e) => q.state[1](e.target.value)}
+      required={q.required ? true : undefined}
+    ></input>
+  );
+}
+
+function renderExtraQuestions(
+  questions: Array<
+    Question & { state: [string, Dispatch<SetStateAction<string>>] }
+  >
+) {
+  return questions.map((q, i) => (
+    <>
+      <div key={`label-${i.toString()}`}>
+        <label>{q.question}</label>
+      </div>
+      <div key={`input-${i.toString()}`}>{renderInputForQuestion(q)}</div>
+    </>
+  ));
+}
+
+export default function PetitionCard(props: Props) {
+  const petition = props.petition;
+  const [state, setState] = useState(State.Initial);
+  useEffect(() => {
+    if (state != State.Success && ensureSessionStorage()[petition.slug]) {
+      setState(State.Existed);
+    }
+  });
+
+  const extraQuestions = (petition.extraQuestions || []).map((q) => ({
+    state: useState(""),
```

Comment:
Ik raad heel sterk af om een useState te gebruiken in een map. Het verbaasd mij dat dit überhaupt kan (zou onmogelijk moeten zijn om hooks op een andere plek dan direct in je functional component body te gebruiken.

Reply:
De hele array moet de state zelf zijn. Dus `const extraQuestions = useState((petition.extraQuestions ?? []).map....)`

Reply:
poeh, okay dan